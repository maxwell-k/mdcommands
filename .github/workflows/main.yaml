# .github/workflows/main.yaml
# Copyright 2025 Keith Maxwell
# SPDX-License-Identifier: CC0-1.0

on: # yamllint disable-line rule:truthy
  push: { branches: [main] }
  pull_request: { branches: [main] }
  workflow_dispatch:

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5
      - uses: denoland/setup-deno@v2
      - run: deno --version
      - run: deno task all
      - run: git diff --exit-code # changes cause this job to fail
      # yamllint disable-line rule:line-length
      - uses: actions/upload-artifact@ea165f8d65b6e75b540449e92b4886f43607fa02 # v4
        with: { path: coverage/html/ }
      - uses: astral-sh/setup-uv@557e51de59eb14aaaba2ed9621916900a91d50c6 # v6
        with: { enable-cache: false }
      - run: uv tool run reuse lint
  renovate-config-validator:
    # yamllint disable-line rule:line-length
    uses: maxwell-k/dotlocalslashbin/.github/workflows/renovate.yaml@cb60f00144f887cd482609b344ae8ed78feadeed # v0.0.19
  commitlint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5
        with: { fetch-depth: 0 }
      - run:
          podman run "--volume=$PWD:/src:z" --workdir=/src commitlint/commitlint
          --verbose --from=origin/main --to=HEAD
