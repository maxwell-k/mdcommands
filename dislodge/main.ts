// dislodge/main.ts
// SPDX-License-Identifier: MPL-2.0
// Copyright 2025 Keith Maxwell

import { marked, Token } from "npm:marked@^16.3.0";
import { EOL } from "jsr:@std/fs@^1.0.19";
import { dirname, join } from "jsr:@std/path@^1.1.2";

// https://github.com/zakhenry/embedme/blob/master/src/embedme.lib.ts#L458
const marker = /<!--\s*?embedme[ ]+?(\S+?)\s*?-->/;

export function main(args: string[]): string {
  const [file] = args.length ? args : ["README.md"];
  const markdown: string = Deno.readTextFileSync(file);

  type Block = [
    path: string,
    data: string,
  ];
  const blocks: Block[] = [];

  let path: string | null = null;
  const walkTokens = (token: Token) => {
    if (path && token.type === "code") {
      blocks.push([path, token.text + "\n"]);
      path = null;
    } else if (token.type === "html") {
      const match = token.raw.match(marker);
      if (match) path = join(dirname(file), match[1]);
    }
  };

  marked.use({ walkTokens });
  marked.parse(markdown);

  for (const block of blocks) Deno.writeTextFileSync(...block);
  return blocks.map(([path, data]) =>
    `Wrote ${data.length} characters to ${path}`
  ).join(EOL);
}

// deno-coverage-ignore
if (import.meta.main) console.log(main(Deno.args));
