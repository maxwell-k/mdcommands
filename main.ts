// main.ts
// SPDX-License-Identifier: MPL-2.0
// Copyright 2024 Keith Maxwell

import { marked, Token } from "npm:marked@16";
import { EOL } from "jsr:@std/fs/eol";

export function mdcommands(markdown: string): string[] {
  const walkTokens = (token: Token) => {
    if (token.type === "code") {
      const text = token.text.replaceAll("\\\n", "").replaceAll(/ +/g, " ");
      result.push(...text.split(/\r?\n/));
    }
  };

  const result: string[] = [];
  marked.use({ walkTokens });
  marked.parse(markdown);
  return result;
}

export function main(args: string[]): string {
  const [file] = args.length ? args : ["README.md"];
  const result: string[] = mdcommands(Deno.readTextFileSync(file));
  return result.join(EOL);
}

if (import.meta.main) console.log(main(Deno.args));
