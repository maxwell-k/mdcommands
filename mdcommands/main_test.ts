// mdcommands/mdcommands_test.ts
// SPDX-License-Identifier: MPL-2.0
// Copyright 2024 Keith Maxwell
import { main, mdcommands } from "./main.ts";
import { assertEquals } from "jsr:@std/assert@^1.0.14";
import { assertSnapshot } from "jsr:@std/testing@^1.0.15/snapshot";
import { join } from "jsr:@std/path@^1.1.2";

Deno.test("indented code block test", () => {
  const result = mdcommands(`
Paragraph:

    indented code block
`);
  assertEquals(result, ["indented code block"]);
});

Deno.test("fenced code block test", () => {
  const result = mdcommands(`
Paragraph:

\`\`\`
fenced code block
\`\`\`
`);
  assertEquals(result, ["fenced code block"]);
});

Deno.test("joins lines that use the backslash continuation character", () => {
  const result = mdcommands(`
Paragraph:


    echo \
      1 \
      2 \
      3
    echo 4

`);
  assertEquals(result, ["echo 1 2 3", "echo 4"]);
});

Deno.test(
  `snapshot test against example.md`,
  async function (t): Promise<void> {
    const result: string = main([
      join(import.meta.dirname as string, "example.md"),
    ]);
    await assertSnapshot<string>(t, result);
  },
);

Deno.test(
  `snapshot test against defaults`,
  async function (t): Promise<void> {
    const result: string = main([]);
    if (Deno.cwd() == import.meta.dirname) {
      await assertSnapshot<string>(t, result);
    } else { // to allow running tests from the repository root
      assertEquals(result, "");
    }
  },
);
