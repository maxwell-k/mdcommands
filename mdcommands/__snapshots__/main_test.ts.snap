export const snapshot = {};

snapshot[`snapshot test against example.md 1`] = `
"deno --version
fzf --version
echo 1 2 3"
`;

snapshot[`snapshot test against defaults 1`] = `
"deno install --global --name mdcommands --allow-read https://raw.githubusercontent.com/maxwell-k/mdcommands/main/mdcommands/main.ts
mdcommands example.md
mdcommand example.md"
`;
